### PR TITLE
Fixing https://github.com/wso2/product-apim/issues/3147

### PR DESCRIPTION
**Pull request by ChamNDeSilva**
### Proposed changes in this pull request

Fixing https://github.com/wso2/product-apim/issues/3147
-

### When should this PR be merged

ASAP

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java [hunk 1]
```diff
@@ -400,6 +400,12 @@
     //Property for enabling scope sharing between APIs
     public static final String ENABLE_API_SCOPES_SHARING = "enable-api-scopes-sharing";
 
+    // Constants for obtaining organization claims
+    public static final String CHECK_ORGANIZATION_FROM_SAML_ASSERTION = "checkOrganizationFromSamlAssertion";
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java]: shall we rename this to "readOrganizationClaimFromSamlAssertion"? (note read and claim parts.)

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java [hunk 2]
```diff
@@ -400,6 +400,12 @@
...
+    public static final String SAML2_SSO_AUTHENTICATOR_NAME = "SAML2SSOAuthenticator";
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java]: Shouldn't this be "SAMLSSOAuthenticator"? IINM that's what we have in application-authentication.xml.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java [hunk 3]
```diff
@@ -400,6 +400,12 @@
...
+    public static final String ORGANIZATION_CLAIM_ATTRIBUTE = "OrganizationClaimAttribute";
+    public static final String ORGANIZATION_ATTRIBUTE_NAME = "http://wso2.org/claims/organization";
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java]: Shall we rename this to "DEFAULT_ORGANIZATION_CLAIM_NAME"?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/SAMLGroupIDExtractorImpl.java [hunk 4]
```diff
@@ -79,15 +86,21 @@ public String getGroupingIdentifiers(String loginResponse) {
...
-                    .getTenantId(tenantDomain);
-            UserRealm realm = (UserRealm) realmService.getTenantUserRealm(tenantId);
-            UserStoreManager manager = realm.getUserStoreManager();
-            organization =
-                    manager.getUserClaimValue(MultitenantUtils.getTenantAwareUsername(username), claim, null);
+            String isSAML2Enabled = System.getProperty(APIConstants.CHECK_ORGANIZATION_FROM_SAML_ASSERTION);
+
+            if (!StringUtils.isEmpty(isSAML2Enabled) && Boolean.parseBoolean(isSAML2Enabled)) {
+                organization = getOrganizationFromAssertion(assertions);
+
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/SAMLGroupIDExtractorImpl.java]: extra new line

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/SAMLGroupIDExtractorImpl.java [hunk 5]
```diff
@@ -123,6 +136,101 @@ public String getGroupingIdentifiers(String loginResponse) {
...
+        return APIConstants.ORGANIZATION_ATTRIBUTE_NAME;
+    }
+
+    /**
+     * Get the organization list from the SAML2 Assertion
+     *
+     * @param assertions SAML2 assertions returned in SAML response
+     * @return Organization list from the assertion
+     */
+    private String getOrganizationFromAssertion(List<Assertion> assertions) {
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/SAMLGroupIDExtractorImpl.java]: let's rename to SamlAssersion.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java (hunk 2 above)
Comment (ChamNDeSilva) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java]: No. If it is SAMLSSOAuthenticator, it refers to the application-authenticators.xml file in <AM_HOME>/repository/identity directory.
The correct file to be refered is authenticators.xml file in  <AM_HOME>/repository/conf/security directory. 
Where we can find following configuration.

<Authenticator name="SAML2SSOAuthenticator" disabled="false">
        <Priority>0</Priority>
        <Config>
            <Parameter name="LoginPage">/carbon/admin/login.jsp</Parameter>
            <Parameter name="ServiceProviderID">carbonServer</Parameter>
            <Parameter name="IdentityProviderSSOServiceURL">https://localhost:9444/samlsso</Parameter>
            <Parameter name="NameIDPolicyFormat">urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified</Parameter>
            <Parameter name="AssertionConsumerServiceURL">https://localhost:9443/acs</Parameter>
            <Parameter name="AssertionSignatureValidationEnabled">true</Parameter>
	    <Parameter name="OrganizationClaimAttribute">http://wso2.org/claims/organization</Parameter>
        </Config>

**Pull request merged**